File: 2022/day02.cpp
// https://adventofcode.com/2022/day/2

#include "../common/common.h"
#include "../common/rust.h"

#include <print>
#include <ranges>
#include <span>
#include <string_view>

auto parse(String const& filename) -> Vec<std::pair<char, char>> {
  return aoc::views::read_lines(filename) |
         stdv::transform([](std::string_view line) {
           auto [a, b] = aoc::split_once(line, ' ');
           return std::pair{a.at(0), b.at(0)};
         }) |
         aoc::collect_vec<std::pair<char, char>>();
}

enum hand {
  // Starts at 0 instead of 1 so we can do modulo 3
  // Need to add 1 to score accordingly
  rock = 0,
  paper = 1,
  scissors = 2,
};

enum outcome {
  lose = 0,
  draw = 3,
  win = 6,
};

fn solve_case1(std::span<const std::pair<char, char>> input) -> int {
  int score = 0;

  aoc::flat_map<char, int> hand_map = {
      {'A', rock}, {'B', paper}, {'C', scissors},
      {'X', rock}, {'Y', paper}, {'Z', scissors},
  };

  for (auto [opponent_char, response_char] : input) {
    let opponent = hand_map.at(opponent_char);
    let response = hand_map.at(response_char);
    // Add 1 because values are lower
    score += response + 1;
    if (opponent == response) {
      score += draw;
      continue;
    }
    if (((response == rock) && (opponent == scissors)) ||
        ((response - opponent) == 1)) {
      score += win;
    } else {
      score += lose;
    }
  }

  return score;
}

fn solve_case2(std::span<const std::pair<char, char>> input) -> int {
  int score = 0;

  aoc::flat_map<char, int> hand_map = {
      {'A', rock}, {'B', paper}, {'C', scissors},
      {'X', lose}, {'Y', draw},  {'Z', win},
  };

  for (auto [opponent_char, outcome_char] : input) {
    let opponent = hand_map.at(opponent_char);
    let outcome = hand_map.at(outcome_char);
    score += outcome;
    if (outcome == draw) {
      // Add 1 because values are lower
      score += opponent + 1;
      continue;
    }
    int response = 0;
    if (outcome == win) {
      response = ((opponent + 1) % 3);
    } else {
      response = ((3 + opponent - 1) % 3);
    }
    // Add 1 because values are lower
    score += response + 1;
  }

  return score;
}

int main() {
  std::println("Part 1");
  let example = parse("day02.example");
  AOC_EXPECT_RESULT(15, solve_case1(example));
  let input = parse("day02.input");
  AOC_EXPECT_RESULT(14375, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(12, solve_case2(example));
  AOC_EXPECT_RESULT(10274, solve_case2(input));
  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day02.rs
type Input = Vec<(u8, u8)>;

fn parse(filename: &str) -> Input {
    aoc::file::read_lines(filename)
        .into_iter()
        .map(|line| {
            let b = line.into_bytes();
            (b[0], b[2])
        })
        .collect()
}

fn solve_case1(input: &Input) -> i32 {
    // Hands: rock=0, paper=1, scissors=2
    input
        .iter()
        .map(|(opp, resp)| {
            let opponent = (opp - b'A') as i32;
            let response = (resp - b'X') as i32;
            // Add 1 because hand values are zero-based
            let mut score = response + 1;
            if opponent == response {
                score += 3;
            } else if (response == 0 && opponent == 2) || (response - opponent == 1) {
                score += 6;
            }
            score
        })
        .sum()
}

fn solve_case2(input: &Input) -> i32 {
    // Outcomes: lose=0, draw=3, win=6
    input
        .iter()
        .map(|(opp, out)| {
            let opponent = (opp - b'A') as i32;
            let outcome = match out {
                b'X' => 0,
                b'Y' => 3,
                b'Z' => 6,
                _ => unreachable!(),
            };
            let mut score = outcome;
            if outcome == 3 {
                // Add 1 because hand values are zero-based
                score += opponent + 1;
            } else {
                let response = if outcome == 6 {
                    (opponent + 1) % 3
                } else {
                    (3 + opponent - 1) % 3
                };
                // Add 1 because hand values are zero-based
                score += response + 1;
            }
            score
        })
        .sum()
}

fn main() {
    println!("Part 1");
    let example = parse("day02.example");
    aoc::expect_result!(15, solve_case1(&example));
    let input = parse("day02.input");
    aoc::expect_result!(14375, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(12, solve_case2(&example));
    aoc::expect_result!(10274, solve_case2(&input));
}
